tests/5.0/tasks/test_task_depend_iterator.cpp

```diff
@@ -0,0 +1,56 @@
+//===-- test_task_depend_iterator.cpp - test of the iterator clause inside the task depend clause ----===//
+// 
+// OpenMP API Version 5.0 Nov 2020
+// 
+// This file is a test for the iterator modifier when used with the task depend
+// clause. This modifier should create an iterator that expand into multiple values
+// inside the clause they appear. In this particular test case the iterator expands into
+// several values creating several dependencies at the same time.
+//
+//===-------------------------------------------------------------------------------------------------===//
+
+#include <omp.h>
+#include <algorithm>
+#include <chrono>
+#include <cstdlib>
+#include <iostream>
+#include <thread>
+#include <vector>
+#include "ompvv.h"
+
+int test_task_depend_iterator() {
+  int ptr[] = {0, 4, 5, 6, 7, 8, 9, 10, 11}, cols[] = {1, 2, 3, 4, 5, 5, 6, 6, 7, 7, 8};
+  std::vector<int> threadOrder;
+  bool threadOrderError = false;
+#pragma omp parallel num_threads(4)
+  {
+#pragma omp single
+    {
+      for (int i = 0; i < 8; ++i) {
+        int pos = ptr[i], size = ptr[i + 1] - ptr[i];
+#pragma omp task depend(iterator(it = 0 : size), in : ptr[cols[pos + it]]) depend(out : ptr[i])
```

Comment:
The sintax according to OpenMP 5.0 specification is:
```
iterator(iterators-definition)
```
where `iterators-definition` is:

```
iterator-specifier [,iterators-definition ]
```

And `iterator-specifier` is:
```
[ iterator-type ]  identifier=range-specification
```

And `iterator-type` is a type name.

It seems to me that you will need to define it with a typename. I will check it out, but it should be something like:

```
iterator(std::vector<int>::iterator  it = 0 : size)
```

In this case the type is: `std::vector<int>::iterator` , the identifier is `it` and the range is `0:size`

Reply:
also in `ptr[cols[pos + it]]` the iterator must be dereference:
`ptr[cols[pos + *it ]]`

Reply:
Uhmmm actually, the range has to also be an iterator object:


`range-specification` is of the form `begin:end[:step]`, where `begin` and `end` are **expressions for which their types can be converted to** `iterator-type` and `step` is an integral expression.
```

Reply:
As far I've seen and read, the iterator clause has no relation with C++ std iterators and OpenMP iterators are integer numbers defined by specification. If you check the OpenMP examples for 5.0, the clause comes without a type. That's why you don't have to de-reference the iterator.

Reply:
Actually, the iterator must be of type integer or pointer according to the restrictions:

```
The iterator-type must be an integral or pointer type.
```

I made this change and now it compiles and run with GCC 9.2

```
#pragma omp task depend(iterator(int it = 0 : size), in : ptr[cols[pos + it]]) depend(out : ptr[i])
```

Uhmm, I will bring this up with the spec people to see what they have to say

Reply:
From Deepak:

> The brackets indicate that the iterator-type is optional. The spec also says that if it is missing, the type is implicitly int for C/C++ and the default integer for Fortran.

> I think (iterator-type) would (be needed) if the expressions used in the range specification are pointer types

So this test should be valid.